File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
input = sys.stdin.readline
from itertools import combinations

n = int(input())

arr = []
result = sys.maxsize
for _ in range(n):
    arr.append(list(map(int, input().split())))

l = [i for i in range(n)]
a = []
for v in combinations(l, n//2):
    a.append(v)

for i in range(len(a)//2):
    x=0
    y=0
    for q in combinations(a[i], 2):
        x+=arr[q[0]][q[1]]+arr[q[1]][q[0]]

    for w in combinations(a[-(i+1)], 2):
        y+=arr[w[0]][w[1]]+arr[w[1]][w[0]]

    result = min(abs(x-y), result)

print(result)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java


import java.util.*;
import java.io.*;

public class Main {
    static int n; //짝수
    static int[][] list;
    static ArrayList<Integer> result;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        n = Integer.parseInt(br.readLine());
        list = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                list[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        ArrayList<Integer> tmp = new ArrayList<>();
        result = new ArrayList<>();
        recur(tmp, 0);
        System.out.println(Collections.min(result));
    }
    static void recur(ArrayList<Integer> list1, int index){
        if (list1.size() == n / 2) {
            int sum = 0;
            for (int i = 0; i < n - 1; i++) {
                for (int j = i+1; j < n;j++) {
                    if ((list1.contains(i)) && (list1.contains(j))){
                        sum+=(list[i][j]+list[j][i]);
                    }
                    else if((!list1.contains(i)) && (!list1.contains(j))){
                        sum-=(list[i][j]+list[j][i]);
                    }
                }
            }
            result.add(Math.abs(sum));
            return;
        }
        for (int i = index; i < n; i++) {
            list1.add(i);
            recur(list1, i+1);
            list1.remove(list1.size()-1);
        }
    }
}
